File: LeetCode/590.py
#Link: https://leetcode.com/problems/n-ary-tree-postorder-traversal/

"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def postorder(self, root: 'Node') -> List[int]:
        res = []
        stack = deque()
        stack.append(root)
        while stack:
            temp = stack.pop()
            if temp:
                res.append(temp.val)
                for child in temp.children:
                    stack.append(child)
        return reversed(res)
        

File: LeetCode/cpp/590.cpp
//Link: https://leetcode.com/problems/n-ary-tree-postorder-traversal/

/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    vector<int> postorder(Node* root) {
        Node* temp;
        vector<int> res;
        stack<Node*> s;
        s.push(root);
        while(!s.empty()) {
            temp = s.top();
            s.pop();
            if(temp) {
                res.push_back(temp->val);
                for(auto child : temp->children)
                    s.push(child);
            }
        }
        reverse(res.begin(),res.end());
        return res;
    }
};
